torch/_dynamo/variables/torch.py

```diff
@@ -976,6 +985,93 @@ def call_function(
         from . import ConstantVariable, SymNodeVariable, TensorVariable
         from .builder import wrap_fx_proxy
 
+        if self.traceable:
+            import torch._higher_order_ops.flat_apply as flat_apply
+            from torch._higher_order_ops.flat_apply import (
+                func_to_graphable,
+                is_graphable_type,
+            )
+            from torch.utils._pytree import tree_flatten
+
+            fn = self.value
+            # 1. Convert `args, kwargs` into pytree-flattened proxy forms.
+            #
+            # Rather than reconstructing `args, kwargs` into python objects and
+            # then tree_flatten them, we just let Dynamo symbolically interpret
+            # `tree_flatten((args, kwargs))`. This saves us from having to
+            # worry about the reconstruction logic, side effects, and guards.
+            packed_input_vt = TupleVariable.build(
+                tx, (TupleVariable.build(tx, args), ConstDictVariable.build(tx, kwargs))
+            )
+            out_vt = variables.UserFunctionVariable(tree_flatten).call_function(
+                tx, [packed_input_vt], {}
+            )
+            assert isinstance(out_vt, TupleVariable) and len(out_vt.items) == 2
+            flat_args_vts, input_spec_vt = out_vt.items
+            assert isinstance(flat_args_vts, ListVariable)
+
+            # Handle the case when the input contains a non-graphable type.
+            for flat_arg_vt in flat_args_vts.items:
+                arg_type = flat_arg_vt.python_type()
+                if not is_graphable_type(arg_type):
+                    unimplemented(
+                        f"""
+Attempting to call a `mark_traceable`-ed function with arguments that contain a value of type <{flat_arg_vt.python_type().__qualname__}>, please use one of the following to register the type with pytree:
+  * `torch.utils._pytree.register_pytree_node`
+"""  # NOQA: B950
+                    )
+
+            # Since we checked with `is_graphable` above, `as_proxy` on the
+            # flat_arg VT should always work.
+            proxified_flat_args = [
+                flat_arg_vt.as_proxy() for flat_arg_vt in flat_args_vts.items
+            ]
+
+            # The downstream `flat_apply` call requires the input spec; however,
+            # it's not a graphable type. So we still have to reconstruct it into
+            # a python object, and store it as a constant attribute on the fx
+            # graph.
+            #
+            # Luckily, the structure of input spec is predictable and fixed
+            # (it's a part of the torch codebase), and existing
+            # `VariableTracker.as_python_constant` implementations already
+            # covers all the cases, so we just use that.
```

Comment:
this is not completely true, someone might have a user-defined pytree where they toss some type that doesn't have an as_python_constant into the TreeSpec. This happens in the register_constant case, but can also happen in the other cases too

Reply:
> but can also happen in the other cases too

Do you mind elaborating on those? Would the try/except block below account for that?

Reply:
The try/except block accounts for that. I am pointing out that the comment here is not precise.

We can have a custom user-defined pytree type where tree_flattening-it stashes something into the TreeSpec that is not reconstructible via as_python_constant. An example:
```
class Point:
    x: Tensor
    y: Tensor
    z: weird_object

def flatten_fn(p):
    # first return is the leaves, second return is the context that goes into the TreeSpec
    return [p.x, p.y], weird_object

register_pytree_node(Point, flatten_fn, unflatten_fn)
```